### PR TITLE
Implemented Order 

### PR DESCRIPTION
**Pull request by holsoni**

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        return orderService.getOrderHistory(authentication, pageable);
+    }
+
+    @PatchMapping("/{id}")
+    @Operation(summary = "Update order status",
+            description = "Update the status of an existing order")
+    public void updateOrderStatus(@PathVariable Long id,
+                                  @RequestBody UpdateOrderStatusRequest requestDto,
+                                  Authentication authentication) {
+        orderService.updateOrderStatus(id, requestDto, authentication);
```
Comment (CodeWebWeaver) [inline src/main/java/com/example/bookstore/controller/OrderController.java]: You shoudn`t use authentication here
You need PreAuthorize User

File: src/main/java/com/example/bookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                                            Pageable pageable) {
+        return orderService.getOrderItems(orderId, pageable);
+    }
+
+    @GetMapping("/{orderId}/items/{itemId}")
+    @Operation(summary = "Get order item by ID",
+            description = "Get a specific order item by its ID")
+    public OrderItemDto getOrderItemById(@PathVariable Long orderId,
+                                         @PathVariable Long itemId) {
+        return orderService.getByIdAndOrderId(orderId, itemId);
```
Comment (CodeWebWeaver) [inline src/main/java/com/example/bookstore/controller/OrderController.java:70]: You need to authorize user here

File: src/main/java/com/example/bookstore/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        orderService.updateOrderStatus(id, requestDto, authentication);
+    }
+
+    @GetMapping("/{orderId}/items")
+    @Operation(summary = "Get order items", description = "Get the items for a specific order")
+    public List<OrderItemDto> getOrderItems(@PathVariable Long orderId,
+                                            Pageable pageable) {
+        return orderService.getOrderItems(orderId, pageable);
+    }
+
```
Comment (CodeWebWeaver) [inline src/main/java/com/example/bookstore/controller/OrderController.java:63]: You need authorize user here

File: src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class OrderServiceImpl implements OrderService {
+    private final ShoppingCartRepository cartRepository;
+    private final ShoppingCartMapper cartMapper;
+    private final OrderRepository orderRepository;
```
Comment (CodeWebWeaver) [inline src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java]: reduncant field

File: src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        Order order = new Order();
+        order.setUser((User)authentication.getPrincipal());
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Status.PROCESSING);
+        order.setShippingAddress(requestDto.getShippingAddress());
+        order.setTotal(countTotal(order.getOrderItems()));
+        order.setOrderItems(getOrderItemsFromCart(shoppingCart, order));
+        orderRepository.save(order);
+        orderItemRepository.saveAll(order.getOrderItems());
+    }
```
Comment (CodeWebWeaver) [inline src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java]: you can split it in method

**Review by CodeWebWeaver — CHANGES_REQUESTED** [review]
You can add additional handlers for exceptions if you want.
Also I can`t find how you clear your shopping cart after placing order

File: src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        return orderRepository
+                .getAllByUserId(getCurrentUserId(authentication), pageable).stream()
+                .map(orderMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void updateOrderStatus(Long id,
+                                  UpdateOrderStatusRequest requestDto,
+                                  Authentication authentication) {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java]: Do you use authentication here?

File: src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    @Override
+    public OrderItemDto getByIdAndOrderId(Long orderId, Long itemId) {
+        return orderItemMapper.toDto(orderItemRepository.findByIdAndOrderId(
+                orderId, itemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Sorry! Can't find order item with id  " + itemId
+                        + ", order id " + orderId)));
+    }
+
+    private Order getOrder(PlaceOrderRequestDto requestDto,
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java]:
```suggestion
    private Order createOrder(PlaceOrderRequestDto requestDto,
```

File: src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                .orElse(BigDecimal.ZERO);
+    }
+
+    private Set<OrderItem> getOrderItemsFromCart(ShoppingCart shoppingCart, Order order) {
+        return shoppingCart.getCartItems().stream()
+                .map(item -> mapCartItemToOrderItem(item, order))
+                .collect(Collectors.toSet());
+    }
+
+    private OrderItem mapCartItemToOrderItem(CartItem cartItem, Order order) {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookstore/service/order/impl/OrderServiceImpl.java]: Is it possible to use mapper for mapping?

File: src/main/resources/db/changelog/changes/13-create-orders-table.yaml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValue: false
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/changes/13-create-orders-table.yaml]: add empty line

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, but some improvements required

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**